### PR TITLE
Support pg 15

### PR DESCRIPTION
**Pull request by lubennikovaav**
- Parameterize functions that depend on PostgreSQL version. Use either explicitly passed `pg_version` or `DEFAULT_PG_VERSION`.
- Refactor `postgres_ffi` code to separate version-specific and version-independent code.
- Adjust safekeepers code to take `pg_version` into account.
- Preserve timeline's `pg_version` in `TimelineMetadata`, to make it available after timeline offloading.
- Update `neon_local`: commands that create timeline (explicitly or implicitly) now have optional `pg-version` parameter.
Usage example:
```
# implicitly creates initial_timeline of version 15
./target/debug/neon_local init --pg-version=15  
./target/debug/neon_local start

# uses initial_timeline of version 15
./target/debug/neon_local pg start main --pg-version=15

# this will fail, because initial timeline has version 15 
./target/debug/neon_local pg start main --pg-version=14
```
TODO:
- [x] Handle TimelineMetadata format change properly. Bump storage version? Implement backward compatibility.
- [ ] Add more tests
- [x] Update openapi_specs
- [x] Parameterize python tests to be able to run them against v15
``` 
usage example is
DEFAULT_PG_VERSION=15 ./scripts/pytest
```

- [ ] fix failing tests on pg_15: (see CI run here https://github.com/neondatabase/neon/pull/2471)
```
ERROR test_runner/regress/test_pg_regress.py::test_isolation - Exception:    ...
FAILED test_runner/regress/test_pg_regress.py::test_pg_regress - AssertionErr...
FAILED test_runner/regress/test_pg_regress.py::test_isolation - subprocess.Ca...
FAILED test_runner/regress/test_pg_regress.py::test_sql_regress - AssertionEr...
FAILED test_runner/regress/test_readonly_node.py::test_timetravel - psycopg2....
FAILED test_runner/regress/test_timeline_size.py::test_timeline_size_metrics
```

### CODE REVIEW

File: pageserver/src/basebackup.rs [hunk 1]
```diff
@@ -269,15 +269,16 @@ where
         };
 
         // TODO pass this as a parameter
```
Comment (hlinnaka) [inline pageserver/src/basebackup.rs:271]:
```suggestion
```

File: pageserver/src/basebackup.rs [hunk 2]
```diff
@@ -269,15 +269,16 @@ where
...
+            let pg_version_str = self.timeline.pg_version.to_string();
+            let header = new_tar_header("PG_VERSION", pg_version_str.len() as u64)?;
+            self.ar.append(&header, pg_version_str.as_bytes())?;
 
-            let header = new_tar_header("global/PG_VERSION", version_bytes.len() as u64)?;
-            self.ar.append(&header, version_bytes)?;
+            info!("timeline.pg_version {}", self.timeline.pg_version);
+
+            let header = new_tar_header("global/PG_VERSION", pg_version_str.len() as u64)?;
+            self.ar.append(&header, pg_version_str.as_bytes())?;
```
Comment (hlinnaka) [inline pageserver/src/basebackup.rs:281]: Not new with this PR, but I just noticed there actually isn't a PG_VERSION file in `global` in PostgreSQL. It doesn't hurt I guess, but we could just remove this.

File: pageserver/src/pgdatadir_mapping.rs [hunk 3]
```diff
@@ -1128,7 +1128,9 @@ static ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; BLCKSZ as usize]);
 //
 // Checkpoint:
 // 03 00000000 00000000 00000000 00   00000001
-
+//
+// PG_VERSION:
+// 03 00000000 00000000 00000000 00   00000002
```
Comment (hlinnaka) [inline pageserver/src/pgdatadir_mapping.rs:1133]: I don't see any code to create this key.

File: pageserver/src/walrecord.rs [hunk 4]
```diff
@@ -3,12 +3,11 @@
 //!
```
Comment (hlinnaka) [inline pageserver/src/walrecord.rs:3]: I've got a feeling that we will have more version-specific differences in this file in the future. This is ok for now, but we'll probably need to refactor it more in the future.

**Review by hlinnaka — COMMENTED** [review]
Great, this seems tantalizingly close to being ready!

File: pageserver/src/pgdatadir_mapping.rs (hunk 3 above)
Comment (lubennikovaav) [reply pageserver/src/pgdatadir_mapping.rs]: fixed

File: libs/postgres_ffi/src/waldecoder.rs [hunk 5]
```diff
@@ -16,55 +17,19 @@ use crc32c::*;
...
-    inputbuf: BytesMut,
-    state: State,
-}
-
-#[derive(Error, Debug, Clone)]
-#[error("{msg} at {lsn}")]
-pub struct WalDecodeError {
-    msg: String,
-    lsn: Lsn,
+pub trait WalStreamDecoderHandler {
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply libs/postgres_ffi/src/waldecoder.rs:22]: I couldn't come up with anything simpler than this implementation, so I've added a comment. 
And also renamed the module to `waldecoder_handler` for clarity.

File: pageserver/src/config.rs [hunk 6]
```diff
@@ -205,7 +206,7 @@ impl Default for PageServerConfigBuilder {
             workdir: Set(PathBuf::new()),
             pg_distrib_dir: Set(env::current_dir()
                 .expect("cannot access current directory")
-                .join("pg_install/v14")),
+                .join(format!("pg_install/v{}", DEFAULT_PG_VERSION))),
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/config.rs:209]: This change requires a lot of refactoring in python tests.
I am working on that now, but I'd rather submit these changes in a follow-up PR.

[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/config.rs]: Fixed. Now we use both PG_DISTRIB_DIR and PG_VERSION to build a path.
Assume that the path follows `PG_DISTRIB_DIR/v{PG_VERSION}` notation.

In rust code, DEFAULT_PG_VERSION remains only in `create_timeline` as a default for API call, and in unit tests.

**Comment by lubennikovaav** [comment]
@MMeent , I fixed everything that you mentioned above and also stabilized the tests.
Anything else?

**Pull request merged**